compiler/rustc_mir/src/transform/inline/cycle.rs

```diff
@@ -0,0 +1,141 @@
+use rustc_data_structures::fx::{FxHashMap, FxHashSet};
+use rustc_data_structures::stack::ensure_sufficient_stack;
+use rustc_hir::def::DefKind;
+use rustc_hir::def_id::{DefId, LocalDefId};
+use rustc_middle::mir::TerminatorKind;
+use rustc_middle::ty::{self, InstanceDef, TyCtxt, subst::SubstsRef};
+
+// FIXME: check whether it is cheaper to precompute the entire call graph instead of invoking
+// this query riddiculously often.
```

Comment:
nit: Not sure if you're planning to leave this in when this is merged, but if so:

```suggestion
// this query ridiculously often.
```